main.go

```diff
@@ -73,11 +74,35 @@ func main() {
 		sigint := make(chan os.Signal, 1)
 		signal.Notify(sigint, os.Interrupt, syscall.SIGTERM)
 		<-sigint
-		s.stop <- struct{}{} // notify having caught signal
+		s.stop <- struct{}{}      // notify having caught signal stop oauthproxy
+		close(metricsStopChannel) // and the metrics endpoint
 	}()
 	s.ListenAndServe()
 }
 
+// startMetricsServer will start the metrics server on the specified address.
+// It always return a channel to signal stop even when it does not run.
+func startMetricsServer(address string) chan struct{} {
+	stop := make(chan struct{}, 1)
+
+	// Attempt to setup the metrics endpoint if we have an address
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
+		go func() {
+			logger.Fatal(s.ListenAndServe())
```

Comment:
Only need to log this if there is an error, or does logger.Fatal check if the error is nil first? 🤔  This is more explicit either way

```suggestion
			if err := s.ListenAndServe(); err != nil {
				logger.Fatal(err)
			}
```

As for whether we should quit if the metrics don't start, I would suggest that's reasonable behaviour, I'd expect if I had requested metrics, I'd want to know if they weren't working

Reply:
`ListenAndServe()` always returns `ErrServerClosed` so there is no need to check for an error. `logger.Fatal(s.ListenAndServe())` is the example used in the `http` package and seems like appropriate usage here. If `ListenAndServe()` stops, our metrics would be unavailable and we wouldn't want to continue running (and if the metrics were monitor we might be signalled to stop anyway).

That being said, we could call `logger.Print` here instead and send a message to the main HTTP server's stop channel. Calling `os.Exit()` as `Fatal` may not be the most graceful approach in retrospect.